File: data/processNwk.py
# recursively peel off outer parentheses
# sibling, (descendants)'nodeName':millionYears
# if !isNumeric(nodeName) -> make a species too

import re

class Node:
  def __init__(self, name: str, parent, yearsSinceParent: int) -> None:
    self.name = name
    self.parent = parent
    self.yearsSinceParent = yearsSinceParent
  def __str__(self) -> str:
    if self.parent == None:
      return f'{self.name}'
    return f'{self.name}, {self.yearsSinceParent} years since {self.parent.name}'
def makeNode(nodeString: str, parentNode):
  # read string -> turn into Node

  # special case: ';\n' is root
  if nodeString == ';\n':
    node = Node('root', None, None)
    isSpecies = False
  else:
    # node without species:   'nodeName':millionYears
    # species node:           Genus_species:millionYears
    if nodeString[0] == '\'':
      isSpecies = True
      pattern = r'\'(\w+)\':([\d\.]+)'
    else:
      isSpecies = False
      pattern = r'([\w_]+):([\d\.]+)'
    match = re.match(pattern, nodeString)
    name = match.group(1).replace('_', ' ')
    years = int(float(match.group(2)) * 1000000)
    node = Node(name, parentNode, years)

  # todo: do something with isSpecies
  print(node)
  return node

def processNwk(nwkPath):
  nwkFile = open(nwkPath, 'r')

  # preprocess
  nwk = nwkFile.read()

  nodeInProgress = ''
  ancestors = []

  print('---------------------------------------------')
  print(nwk)
  print('---------------------------------------------')

  # read backwards
  i = len(nwk) - 1

  while i >= 0:
    char = nwk[i]
    if char == ')':
      # make node AND add to ancestor stack
      ancestor = None if len(ancestors) == 0 else ancestors[-1] # pass None for adding root
      newNode = makeNode(nodeInProgress, ancestor) 
      nodeInProgress = ''
      ancestors.append(newNode)
      i -= 1
      continue
    if char == ',':
      # make new node
      if nodeInProgress:
        makeNode(nodeInProgress, ancestors[-1])
        nodeInProgress = ''
      i -= 1
      continue
    if char == '(':
      # make node AND remove most recent ancestor
      if not nodeInProgress == '':
        # there's a node to make (i.e. not multiple node closures in a row)
        makeNode(nodeInProgress, ancestors[-1])
        nodeInProgress = ''
      ancestors.pop()
      i-= 1
      continue

    nodeInProgress = char + nodeInProgress
    i -= 1


processNwk('/Users/jennyzonka/Code/animle-backend/data/Ursidae_species.nwk')

File: data/processNwk.js
const fs = require('fs/promises');
const { v4: uuidv4 } = require('uuid');

async function makeNode(nodeString, parentNodeId, jsonPath) {
  let pattern;
  let isSpecies = false;
  let name = null;
  let years = null;
  // special case: ';\n' is root
  if (nodeString !== ';\n') {
    // node without species:   'nodeName':millionYears
    // species node:           Genus_species:millionYears
    if (nodeString[0] === `'`) {
      pattern = /\'(\w+)\':([\d\.]+)/;
    } else {
      isSpecies = true;
      pattern = /([\w_]+):([\d\.]+)/;
    }
    // const regex = RegExp(pattern);
    const match = pattern.exec(nodeString);
    name = match[1].replaceAll('_', ' ');
    years = parseFloat(match[2]) * 1000000;
  }
  
  const id = uuidv4();
  const nodeData = {
    _id: id,
    parent: parentNodeId,
    yearsSinceParent: years,
    isSpecies: isSpecies,
    speciesName: isSpecies ? name : null,
    otherNames: isSpecies ? [] : null,
  };
  const content = JSON.stringify(nodeData);
  // write to file
  try {
    await fs.appendFile(jsonPath, content + ',\n');
  } catch (err) {
    console.log(err);
  }

  return id;
}

const processNwk = async (nwkString, jsonPath) => {
  let nodeInProgress = '';
  const ancestors = [];

  // read backwards
  let i = nwkString.length - 1;
  while (i >= 0) {
    const char = nwkString[i];
    switch (char) {
      case ')':
        // make node and add to ancestor stack
        const ancestor = ancestors.length === 0 ? null : ancestors[ancestors.length - 1]; // pass null for adding the root
        const newNode = await makeNode(
          nodeInProgress, ancestor, jsonPath
        );
        nodeInProgress = '';
        ancestors.push(newNode);
        i -= 1;
        break;
      case ',':
        // make new node
        if (nodeInProgress !== '') {
          await makeNode(
            nodeInProgress,
            ancestors[ancestors.length - 1],
            jsonPath
          );
          nodeInProgress = '';
        }
        i -= 1;
        break;
      case '(':
        // make node and remove most recent ancestor
        if (nodeInProgress !== '') {
          // there's a node to make (i.e. not multiple node closures in a row)
          await makeNode(
            nodeInProgress,
            ancestors[ancestors.length - 1],
            jsonPath
          );
          nodeInProgress = '';
        }
        ancestors.pop();
        i -= 1;
        break;
      default:
        nodeInProgress = char + nodeInProgress;
        i -= 1;
    }
  }
}

const processFile = async (nwkPath, jsonPath) => {
  
  // read, parse, and save data
  try {
    // start files and overwrite existing JSON
    await fs.writeFile(jsonPath, '[\n');

    const data = await fs.readFile(nwkPath, {encoding: 'utf8'});

    await processNwk(data, jsonPath);

    // finish files
    await fs.appendFile(jsonPath, ']');
  } catch (err) {
    console.log(err);
  }
}
processFile(
  '/Users/jennyzonka/Code/animle-backend/data/Ursidae_species.nwk',
  '/Users/jennyzonka/Code/animle-backend/data/Ursidae_species_nodes.json',
);

// to seed the databases
// 1. fix the end of the json files manually (facepalm)
// 2. run calculageLineage.js
// 3. use mongoimport
// mongoimport --uri mongodb+srv://jennycade:PASSWORD@cluster0.hhhey.mongodb.net/animle --collection nodes --type json --file /Users/jennyzonka/Code/animle-backend/data/Ursidae_tree.json --jsonArray